pkg/api/write.go

```diff
@@ -15,188 +14,187 @@ import (
-			receivedBatchCount += uint64(len(t.Samples))
+
+		for _, ts := range req.Timeseries {
+			receivedBatchCount += uint64(len(ts.Samples))
 		}
 
-		metrics.ReceivedSamples.Add(float64(receivedBatchCount))
+		m.ReceivedSamples.Add(float64(receivedBatchCount))
 		begin := time.Now()
 
-		numSamples, err := writer.Ingest(timeseries, req)
+		numSamples, err := inserter.Ingest(req.Timeseries, req)
 		if err != nil {
 			log.Warn("msg", "Error sending samples to remote storage", "err", err, "num_samples", numSamples)
 			http.Error(w, err.Error(), http.StatusInternalServerError)
-			metrics.FailedSamples.Add(float64(receivedBatchCount - numSamples))
-			return
+			m.FailedSamples.Add(float64(receivedBatchCount - numSamples))
+			return false
 		}
 
 		duration := time.Since(begin).Seconds()
 
-		metrics.SentSamples.Add(float64(numSamples))
-		metrics.SentBatchDuration.Observe(duration)
+		m.SentSamples.Add(float64(numSamples))
+		m.SentBatchDuration.Observe(duration)
 
-		metrics.WriteThroughput.SetCurrent(getCounterValue(metrics.SentSamples))
+		m.WriteThroughput.SetCurrent(getCounterValue(m.SentSamples))
 
 		select {
-		case d := <-metrics.WriteThroughput.Values:
+		case d := <-m.WriteThroughput.Values:
 			log.Info("msg", "Samples write throughput", "samples/sec", d)
 		default:
 		}
 
-	})
+		return true
+	}
 }
 
-func loadWriteRequest(r *http.Request) (*prompb.WriteRequest, error, string) {
-	var req *prompb.WriteRequest
-
-	switch r.Header.Get("Content-Type") {
-	case "application/x-protobuf":
-		reqBuf, err, msg := decodeSnappyBody(r.Body)
-		if err != nil {
-			return nil, err, msg
-		}
-
-		req = ingestor.NewWriteRequest()
-		if err := proto.Unmarshal(reqBuf, req); err != nil {
-			ingestor.FinishWriteRequest(req)
-			return nil, err, "Protobuf unmarshal error"
-		}
-	case "application/json":
-		var (
-			i    jsonPayload
-			body io.Reader = r.Body
-			err  error
-			msg  string
-		)
-
-		snappyEncoding := strings.Contains(r.Header.Get("Content-Encoding"), "snappy")
-
-		if snappyEncoding {
-			body, err, msg = decodeSnappy(r.Body)
+type writeStage func(http.ResponseWriter, *http.Request, *Metrics) bool
 
-			if err != nil {
-				return nil, err, msg
-			}
-		}
-
-		dec := json.NewDecoder(body)
-		req = ingestor.NewWriteRequest()
-
-		for {
-			if err := dec.Decode(&i); err == io.EOF {
-				break
-			} else if err != nil {
-				ingestor.FinishWriteRequest(req)
-				return nil, err, "JSON decode error"
-			}
-
-			req = appendJSONPayload(req, i)
-		}
-	default:
-		// Cannot get here because of header validation.
-		return nil, fmt.Errorf("unsupported data format (not protobuf or json)"), "Write header validation error"
-	}
-
-	return req, nil, ""
+type writeHandler struct {
+	stages  []writeStage
+	metrics *Metrics
 }
 
-func isWriter(elector *util.Elector) (bool, error) {
-	if elector != nil {
-		shouldWrite, err := elector.IsLeader()
-		return shouldWrite, err
+func (wh *writeHandler) addStages(stages ...writeStage) {
```

Comment:
I think we should make preprocessor as independent package, that has run as a main function in preprocessor interface and stage as individual stage. The run calls all the run() func in the stage.

```golang
type Stage interface {
  Run() (bool, error)
}

type Processor interface {
  AppendStage()
  Run() (bool, error) // Runs all stage.Run().
}
```

Also, writeStage -> stage? Or, better rename stage to step?

Reply:
I would leave this refactor for a different PR and only do it if we refactor both write and read pipelines to use the same interfaces.

Reply:
I am not sure but this can help increase readability. But if you think another PR is fine for this, let's do that way.